Aruco calibration

dfbakin opened the pull request:
PR implemets charuco corners detection to improve calibration precision and robustness in corner cases.

feat: upgraded OpenCV up to 4.8.0 version
feat: charuco detection and calibration
feat: calibration flags

fix: typos, minor problems

dfbakin committed:
charuco modification in progress

dfbakin committed:
debug: tmp version defore container update

dfbakin committed:
feat: charuco calib draft implemented

feat: implemented basic logic of charuco calibration on OpenCV API 4.8.0

fix: deleted redundunt param from CalibrationNode

dfbakin committed:
feat: IoU adapted for arbitrary polygon

feat: visualisation is impleneted for charuco corners

fix: changed detector params to minimise false positives when detecting charuco corners

fix: changed number for iterations for aruco detector for precision reasons

dfbakin committed:
feat: double calibration

feat: changed calibration flags and bad frames elimination for precision reasons

fix: fixed typos in threshold parameter

fix: deleted debug publisher and checked for false positives

dfbakin commented on packages/camera/src/calibration.cpp:
```diff
@@ -124,6 +112,19 @@ CalibrationNode::CalibrationNode() : Node("calibration_node") {
+    param_.charuco_board.setLegacyPattern(true);
+
+    param_.detector_params = cv::aruco::DetectorParameters();
+    param_.detector_params.cornerRefinementMethod = cv::aruco::CORNER_REFINE_APRILTAG;
```
Выбрал этот метод детекции после тестов: необходимо избежать ложных срабатываний при поиске уголков доски. Этот метод был добавлен в OpenCV по [этой статье](https://pcv.oss-cn-shanghai.aliyuncs.com/wp-content/upload/2018/11/wang2016iros.pdf)

dfbakin commented on packages/camera/src/calibration.cpp:
```diff
@@ -124,6 +112,19 @@ CalibrationNode::CalibrationNode() : Node("calibration_node") {
+
+    param_.detector_params = cv::aruco::DetectorParameters();
+    param_.detector_params.cornerRefinementMethod = cv::aruco::CORNER_REFINE_APRILTAG;
+    param_.detector_params.cornerRefinementWinSize = 10;
```
[docs](https://docs.opencv.org/3.4/d1/dcd/structcv_1_1aruco_1_1DetectorParameters.html): `cornerRefinementWinSize: window size for the corner refinement process (in pixels) (default 5).`

dasimagin replied:
Я не вижу смысла это выносить в параметры, коль вы это хардкодите, а потом передаете в конструктор и больше нигде не используете ¯\_(ツ)_/¯

dfbakin commented on packages/camera/src/calibration.cpp:
```diff
@@ -320,23 +334,66 @@ void CalibrationNode::calibrate() {
     CameraIntrinsicParameters intrinsic_params{};
     intrinsic_params.image_size = *state_.frame_size;
+    std::vector<double> per_view_errors;
     double rms = cv::calibrateCamera(
```
Идея в том, чтобы провести калибровка с фиксированными параметрами (точка пересечения главной оптической оси с плоскостью изображения), затем выбросить топ 10% изображений с максимальной ошибкой после первой калибровки
Затем провести повторную калибровку без фиксированных параметров, но опираясь на результат первой калибровки (флаг `cv::CALIB_USE_INTRINSIC_GUESS`)

dfbakin commented on packages/camera/src/calibration.cpp:
```diff
@@ -320,23 +334,66 @@ void CalibrationNode::calibrate() {
+        boost::minmax_element(per_view_errors.begin(), per_view_errors.end());
+    // to eliminate 10% as bad calibration data
+    double border_value =
+        std::max(2.0, *minmax_elem.first + 0.9 * (*minmax_elem.second - *minmax_elem.first));
```
Решил, что не стоит выбрасывать кадры с ошибкой в пределах 2 пикселей, даже если они входят в топ 10% плохих снимков

dfbakin committed:
fix: removed redundant double calibration

dfbakin commented on packages/camera/src/calibration.cpp:
```diff
@@ -320,23 +334,33 @@ void CalibrationNode::calibrate() {
+        cv::noArray(),
+        cv::noArray(),
+        cv::noArray(),
+        per_view_errors,
+        cv::CALIB_FIX_PRINCIPAL_POINT | cv::CALIB_FIX_TAUX_TAUY | cv::CALIB_FIX_S1_S2_S3_S4 |
+            cv::CALIB_ZERO_TANGENT_DIST,
 
+        cv::TermCriteria(cv::TermCriteria::COUNT + cv::TermCriteria::EPS, 50, DBL_EPSILON));
+
```
Опирался на [opencv reference](https://docs.opencv.org/4.8.0/d9/d0c/group__calib3d.html#ga3207604e4b1a1758aa66acb6ed5aa65d)

Считаем, что у нас нет tangetial distortion, при "полной" калибровке получаем околонулевые значения. Также считаем, что центр оптической оси совпадает с центром изображения, что правда при некропнутом снимке

Считаем ограниченный набор коэффициентов дисторсии

dasimagin replied:
> также считаем, что центр оптической оси совпадает с центром изображения

вок так точно опасно делать... может попасться какая-то плохая камера, потратим недели дебага, почему ее сложно откалибровать :)

dfbakin committed:
fix: num of required aruco detections removed

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -2,45 +2,25 @@
  * @return converted point of ROS2 message
  */
-geometry_msgs::msg::Point toMsgPoint(cv::Point2f cv_point) {
+geometry_msgs::msg::Point toMsgPoint(boost::geometry::model::d2::point_xy<double> boost_point) {
```
Хм, можно и про алисик подумать...

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -124,6 +112,19 @@ CalibrationNode::CalibrationNode() : Node("calibration_node") {
+    param_.detector_params.cornerRefinementWinSize = 10;
+    param_.board_params = cv::aruco::CharucoParameters();
+
+    param_.charuco_detector = std::move(std::make_unique<cv::aruco::CharucoDetector>(
+        param_.charuco_board, param_.board_params, param_.detector_params));
```
1. Это не часть параметров, можно просто вынести в ноду...
2. В конкретно этой строке `move` ничего не дает )

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -51,30 +31,38 @@ using namespace boost::geometry;
+ * @param pattern_size width and height of chessboard (in squares)
+ * @return Boost polygon of convex hull containing given corner points
  */
 Polygon convertToBoostPolygon(const std::vector<cv::Point2f>& corners, cv::Size pattern_size) {
```
У вас компилятор тут не ругается, что `pattern_size` больше не используется?

dasimagin replied:
комментарии выше невозможно читать... обычно там пишут имея переменной, а потом описание... как понять `vector all detected charuco corners in cv::Point2f format`? Пройдитесь пожалуйста по комментам.

dasimagin replied:
```@return convex hull of corner points as boost polygon```

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -51,30 +31,38 @@ using namespace boost::geometry;
+ * @param cv_point OpenCV points that will be converted
+ * @return vector of corners, forming a closed polyline of convex hull
+ */
+std::vector<geometry_msgs::msg::Point> toBoardMsgCorners(
+    const std::vector<cv::Point2f>& detected_corners, cv::Size pattern_size) {
```
`patter_size` снова не используется... настройки флажком компиляции во всех пакетах, что такое такие warning-и должны превращаться в ошибки компиляции.

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -172,31 +173,42 @@ void CalibrationNode::handleFrame(const sensor_msgs::msg::CompressedImage::Const
+        signal_.detected_corners->publish(state_.board_corners_array);
+        return;
+    }
+    param_.charuco_board.matchImagePoints(
+        current_corners, current_ids, current_obj_points, current_image_points);
```
Кажется, charuco_board все, что важно стоит хранить в параметрах... даже `pattern_size` больше толком не нужен )

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -51,30 +31,38 @@ using namespace boost::geometry;
+    const std::vector<cv::Point2f>& detected_corners, cv::Size pattern_size) {
+    Polygon poly, hull;
+    for (size_t i = 0; i < detected_corners.size(); ++i) {
+        poly.outer().push_back(Point{detected_corners[i].x, detected_corners[i].y});
```
`emplace_back` не работает?

dasimagin commented on packages/camera/src/calibration.cpp:
```diff
@@ -51,30 +31,38 @@ using namespace boost::geometry;
-    // clang-format on
+    Polygon poly, hull;
+    for (size_t i = 0; i < corners.size(); ++i) {
+        poly.outer().push_back(Point{corners[i].x, corners[i].y});
```
`emplace_back` не работает?

dfbakin committed:
fix: required changes implemented
fix: calibraion.cpp methods' comments corrected
fix: redundunt params eliminated

dfbakin committed:
fix: warnings fixed
fix: CMakeLists.txt treats unused vars as errors
